sw/device/silicon_creator/lib/drivers/flash_ctrl.c

```diff
@@ -99,14 +77,83 @@ rom_error_t flash_ctrl_read_start(uint32_t addr, uint32_t word_count,
   return kErrorOk;
 }
 
-size_t flash_ctrl_fifo_read(size_t word_count, uint32_t *data_out) {
+void flash_ctrl_init(void) {
+  // Initialize the flash controller.
+  abs_mmio_write32(kBase + FLASH_CTRL_INIT_REG_OFFSET,
+                   bitfield_bit32_write(0u, FLASH_CTRL_INIT_VAL_BIT, true));
+}
+
+rom_error_t flash_ctrl_status_get(flash_ctrl_status_t *status) {
+  if (status == NULL) {
+    return kErrorFlashCtrlInvalidArgument;
+  }
+
+  // Read flash operation status.
+  uint32_t op_status = abs_mmio_read32(kBase + FLASH_CTRL_OP_STATUS_REG_OFFSET);
+  // Read flash controller status.
+  uint32_t fc_status = abs_mmio_read32(kBase + FLASH_CTRL_STATUS_REG_OFFSET);
+
+  // Extract operation status bits.
+  status->done = bitfield_bit32_read(op_status, FLASH_CTRL_OP_STATUS_DONE_BIT);
+  status->err = bitfield_bit32_read(op_status, FLASH_CTRL_OP_STATUS_ERR_BIT);
+
+  // Extract flash controller status bits.
+  status->rd_full =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_RD_FULL_BIT);
+  status->rd_empty =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_RD_EMPTY_BIT);
+  status->prog_full =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_PROG_FULL_BIT);
+  status->prog_empty =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_PROG_EMPTY_BIT);
+  status->init_wip =
+      bitfield_bit32_read(fc_status, FLASH_CTRL_STATUS_INIT_WIP_BIT);
+
+  return kErrorOk;
+}
+
+rom_error_t flash_ctrl_read_start(uint32_t addr, uint32_t word_count,
+                                  flash_ctrl_partition_t region) {
+  // Start the read transaction, the value of `erase_type` doesn't matter.
+  return transaction_start(addr, word_count, region, kFlashCtrlErasePage,
+                           FLASH_CTRL_CONTROL_OP_VALUE_READ);
+}
+
+rom_error_t flash_ctrl_prog_start(uint32_t addr, uint32_t word_count,
+                                  flash_ctrl_partition_t region) {
+  // Start the program transaction, the value of `erase_type` doesn't matter.
+  return transaction_start(addr, word_count, region, kFlashCtrlErasePage,
+                           FLASH_CTRL_CONTROL_OP_VALUE_PROG);
+}
```

Comment:
I think we would prefer reads and writes to be blocking, i.e. start the operation, read from/write to the provided buffer, handle the status bits, read, clear, and return the error code.

Please note that a write operation cannot cross program windows. From the [flash_ctrl hwip tech spec](https://docs.opentitan.org/hw/ip/flash_ctrl/doc/#program-resolution):
> Certain flash memories place restrictions on the program window. This means the flash accepts program beats only if all beats belong to the same address window.

Please also take a look at [this function](https://github.com/lowRISC/opentitan/blob/9e18b46e0a96355c3c163629c8a90b70b805f24e/sw/device/lib/flash_ctrl.c#L127) in `sw/device/lib/flash_ctrl.c`